src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/SolutionExplorerVerifierInProcess.cs

```diff
@@ -2,16 +2,78 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Editor;
 using Microsoft.VisualStudio.Extensibility.Testing;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Roslyn.Test.Utilities;
 using Xunit;
 
 namespace Roslyn.VisualStudio.IntegrationTests.InProcess
 {
     [TestService]
     internal partial class SolutionExplorerVerifierInProcess
     {
+        public async Task ActiveDocumentIsSavedAsync(CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            var activeDocument = await TestServices.Editor.GetActiveTextViewAsync(cancellationToken);
+            var editorAdaptersFactoryService = await GetComponentModelServiceAsync<IVsEditorAdaptersFactoryService>(cancellationToken);
+
+            var runningDocumentTable4 = await GetRequiredGlobalServiceAsync<SVsRunningDocumentTable, IVsRunningDocumentTable4>(cancellationToken);
+
+            foreach (var cookie in runningDocumentTable4.GetRunningDocuments())
+            {
+                if (!runningDocumentTable4.IsDocumentInitialized(cookie))
+                    continue;
+
+                var docData = (IVsTextLines)runningDocumentTable4.GetDocumentData(cookie);
+                if (editorAdaptersFactoryService.GetDocumentBuffer(docData) != activeDocument.TextDataModel.DocumentBuffer)
+                    continue;
+
+                runningDocumentTable4.UpdateDirtyState(cookie);
+                if (!runningDocumentTable4.IsDocumentDirty(cookie))
+                {
+                    // We have verified the active document is not dirty
+                    return;
+                }
+
+                throw new InvalidOperationException($"Failed to save the active document: {runningDocumentTable4.GetDocumentMoniker(cookie)}");
+            }
+
+            throw new InvalidOperationException("Failed to locate running document table cookie for the active document");
+        }
+
+        public async Task AllDocumentsAreSavedAsync(CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            var runningDocumentTable = await GetRequiredGlobalServiceAsync<SVsRunningDocumentTable, IVsRunningDocumentTable>(cancellationToken);
+            var runningDocumentTable3 = await GetRequiredGlobalServiceAsync<SVsRunningDocumentTable, IVsRunningDocumentTable3>(cancellationToken);
```

Comment:
Is this better than casting the runningDocumentTable to IVsRunningDocumentTable3?

Reply:
➡️ Simplified to only need `IVsRunningDocumentTable4`